generators/Exercises.fs

```diff
@@ -0,0 +1,448 @@
+        sprintf "let %s = %s" identifier value  
+
+    default this.RenderIdentifier (canonicalDataCase, key, value) = String.camelize key
+
+    default this.RenderIdentifierWithTypeAnnotation (canonicalDataCase, key, value) = 
+        let identifier = this.RenderIdentifier (canonicalDataCase, key, value)
+    
+        match this.IdentifierTypeAnnotation (canonicalDataCase, key, value) with
+        | Some ty -> identifier |> addTypeAnnotation ty
+        | None    -> identifier
+
+    // Canonical-data specific rendering methods
+
+    default this.RenderExpected (canonicalDataCase, key, value) = this.RenderValue (canonicalDataCase, key, value)
+
+    default this.RenderInput (canonicalDataCase, key, value) = this.RenderValue (canonicalDataCase, key, value)
+
+    default this.RenderArrange canonicalDataCase =
+        let renderArrangeProperty property: string option = 
+            match Map.tryFind property canonicalDataCase.Properties with
+            | None -> None
+            | Some value -> Some (this.RenderValueWithIdentifier (canonicalDataCase, property, value))
+
+        canonicalDataCase
+        |> this.PropertiesWithIdentifier 
+        |> List.choose renderArrangeProperty
+
+    default this.RenderAssert canonicalDataCase = 
+        let template = this.ToTestMethodBodyAssertTemplate canonicalDataCase                
+
+        canonicalDataCase
+        |> this.ToTestMethodBodyAssert
+        |> renderPartialTemplate template    
+
+    default this.RenderSut canonicalDataCase = 
+        let parameters = this.RenderSutParameters canonicalDataCase
+        let property = this.RenderSutProperty canonicalDataCase
+        property :: parameters |> String.concat " "
+
+    default this.RenderSutParameters canonicalDataCase =
+        let sutParameterProperties = this.PropertiesUsedAsSutParameter canonicalDataCase
+
+        canonicalDataCase.Properties
+        |> Map.filter (fun key value -> List.contains key sutParameterProperties)
+        |> Map.fold (fun acc key value  -> this.RenderSutParameter (canonicalDataCase, key, value) :: acc) [] 
+        |> List.rev
+    
+    default this.RenderSutParameter (canonicalDataCase, key, value) =
+        this.RenderValueOrIdentifier (canonicalDataCase, key, value) 
+
+    default this.RenderSutProperty canonicalDataCase = string canonicalDataCase.Property
+
+    default this.PropertiesUsedAsSutParameter canonicalDataCase = 
+        canonicalDataCase.Properties
+        |> Map.toList
+        |> List.map fst
+        |> List.except ["property"; "expected"; "description"]
+    
+    // Utility methods to customize rendered output
+
+    default this.PropertiesWithIdentifier canonicalDataCase = []
+
+    default this.IdentifierTypeAnnotation (canonicalDataCase, key, value ) = None
+
+    default this.UseFullMethodName canonicalDataCase = false
+
+    default this.AdditionalNamespaces = []
+
+type Acronym() =
```

Comment:
What's the reason for having all of the exercises in the same file as opposed to each one in a separate file?

Reply:
As far as I can tell, having them all in one file is more idiomatic than have each one in a separate file. In general, as far as I can tell, each file corresponds to a single module, which may include many classes.

However, it may be better to split the `Exercise` class from its implementations. I'll try it to see how it works out.

Reply:
That's really surprising that it's idiomatic because I feel like having them all in the same file goes against isolation. We don't have a lot of people concurrently working on these generators normally, but I feel like if we did having them in the same file would lead to more merge conflicts.

Reply:
@robkeim Well, as in F# you have to specify the dependencies of files in the right order, having lots of small files becomes a bit more tedious than in C#. Here are some examples I found:

- https://github.com/fsharp/FAKE/blob/master/src/app/Fake.Core.Target/Target.fs
- https://github.com/fsharp/FSharp.Compiler.Service/blob/master/src/fsharp/CompileOps.fs
- https://github.com/fsprojects/FSharpLint/blob/master/src/FSharpLint.Core/Framework/Ast.fs
- https://github.com/fsprojects/Paket/blob/master/src/Paket.Core/Common/Domain.fs

I see your point about merge conflicts, but as long as not too many people are working on the same exercise, we'll probably be fine I guess.

The next step is to create individual issues for the test generators I think, just like we did for the C# track.

Reply:
It's been too long since I've done F# I'm definitely getting rusty, thanks for the clarification @ErikSchierboom!